### PR TITLE
Fix 1358 base decode() (0.27->master)

### PR DESCRIPTION
**Pull request by clanmills**
See #1358

I don't know anything about the function base64decode() in src/futils.cpp.  It clearly clips off the last byte (as reported).  I found some code on the internet (URL in the code) to replace the function.  We should review the function base64encode() and adopt code with the same parentage.

We should add a test coverage for both encode and decode functions.  @LeoHsiao1 Could you do this in tests/bashtests/system_tests() and add it to this PR?    You'll need a C++ "driver" as I don't believe there is a simple way to test base64encode() and base64decode() from the exiv2 command line application.  You can add something to samples/remotetest.cpp.  We might not bother to exercise that from the "old" bash tests as they are heading for the firing squad before the next release.  If you need assistance, please ask.

### CODE REVIEW

File: src/futils.cpp
Comment (piponazo) [inline src/futils.cpp:227]: Two comments here. 1st one: we should not change the function name because it is exported in the file `include/exiv2/futils.hpp`:

```cpp
    EXIV2API int base64encode(const void* data_buf, size_t dataLength, char* result, size_t resultSize);
    EXIV2API long base64decode(const char* in, char* out, size_t out_size);
```

Such a change would break the API for exiv2 consumers. We should not declare those functions `static` either.

The 2nd comment is about testing. This is a perfect candidate to be tested with UNIT TESTS (using gtest) instead of system tests (python). I could write the unit tests if you want 😉

**Review by piponazo — CHANGES_REQUESTED** [review]

File: src/futils.cpp
Comment (clanmills) [reply src/futils.cpp:227]: I didn't change the API.  I changed the code in base64encode() which is in the API.

Adding something to the unit_tests.  Great Idea.  Go for that.  Perfect place to test that.

**Comment by clanmills** [comment]
API in fix_1358_base64decode_0.27

```bash
673 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ git status
On branch fix_1358_base64decode_0.27
nothing to commit, working tree clean
674 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ nm -g --demangle lib/libexiv2.dylib | grep base64
00000000000701f0 T Exiv2::base64decode(char const*, char*, unsigned long)
000000000006fe20 T Exiv2::base64encode(void const*, unsigned long, char*, unsigned long)
675 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $
```
API in 0.27-maintenance:
```bash
677 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ git status
On branch 0.27-maintenance
Your branch is up to date with 'origin/0.27-maintenance'.

nothing to commit, working tree clean
678 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ nm -g --demangle lib/libexiv2.dylib | grep base64
0000000000070420 T Exiv2::base64decode(char const*, char*, unsigned long)
0000000000070050 T Exiv2::base64encode(void const*, unsigned long, char*, unsigned long)
679 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ 
```

File: src/futils.cpp
Comment (piponazo) [reply src/futils.cpp:227]: Right, I thought you changed the name of the function but just added a subfunction here. Sorry for the mistake.

Comment (clanmills) [reply src/futils.cpp:227]: Thank You for reviewing this so quickly.

**Comment by clanmills** [comment]
What's going on, Mr Git?  Why have your merged and closed this PR when I only updated it.

I tested this by locally modifying samples/exifprint.cpp as follows:

```cpp
int _tmain(int /*argc*/, _tchar** const /*argv*/)
try {
    Exiv2::XmpParser::initialize();
    ::atexit(Exiv2::XmpParser::terminate);

    char in[1000];
    memset(in,0,sizeof(in));
    for ( int i = 0 ; i < 256 ; i++ ) {
        in[i]= (char)i;
    }
    char out[1000];
    memset(out,0,sizeof(out));

    if ( Exiv2::base64encode(in,256,out,sizeof(out)) == 1 ) {
        std::cout << out << std::endl;
    }
    return 0;
}
//catch (std::exception& e) {
//catch (Exiv2::AnyError& e) {
catch (Exiv2::Error& e) {
    std::cout << "Caught Exiv2 exception '" << e.what() << "'\n";
    return -1;
}
```

This reports exactly the same on the "old" and new implementation of Exiv2::base64encode() as follows:

```bash
703 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ bin/exifprint  
AAECAwQFBgcICQoLDA0ODxAREhMUFRYXGBkaGxwdHh8gISIjJCUmJygpKissLS4vMDEyMzQ1Njc4OTo7PD0+P0BBQkNERUZHSElKS0xNTk9QUVJTVFVWV1hZWltcXV5fYGFiY2RlZmdoaWprbG1ub3BxcnN0dXZ3eHl6e3x9fn+AgYKDhIWGh4iJiouMjY6PkJGSk5SVlpeYmZqbnJ2en6ChoqOkpaanqKmqq6ytrq+wsbKztLW2t7i5uru8vb6/wMHCw8TFxsfIycrLzM3Oz9DR0tPU1dbX2Nna29zd3t/g4eLj5OXm5+jp6uvs7e7v8PHy8/T19vf4+fr7/P3+/w==
704 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/build $ 
```

@piponazo Can you:
1. Review the code changes in src/futils.cpp
2. Should we disable mergify bot.  I don't know what he's doing.

Looking forward to seeing you unit_test for both base64encode() and base64_decode().

**Comment by piponazo** [comment]
It does not look to me like mergify is merging the branch, but just deleting it after it detected the merge. I do not know how this got merged ... I have made the restrictions on 0.27-maintenance a bit harder, by requiring at least 1 approval before the "Merge" action can be triggered.

I also created PR #1361 to revert this change (Just by applyint git revert). I did this because after merging this, the windows build broke:
https://ci.appveyor.com/project/piponazo/exiv2-wutfp/builds/35771699

Once we merge #1361, we can re-open this branch and add the missing tests

**Comment by clanmills** [comment]
Thanks for this feedback, Luis.  Very puzzled that #1360 was merged.  And you're saying that because it was merged, the bot closed it (which makes sense).

I see #1361 is also failing in Visual Studio.  I'll investigate.  It's nothing much, just a size_t/int mismatch or something.

Let's get the build steady.  The unit_tests come later.  I intended to get everything into #1360.  Some plans don't work out!

**Comment by piponazo** [comment]
I know what's happening, I did not do the git revert properly. I'll modify #1361 in the next minutes

**Pull request merged**
